### PR TITLE
bstring

### PR DESCRIPTION
**Pull request by glennsl**
Status:

- [x] Use `camelCase` instead of `snake_case`
- [x] Get rid of scary monadic and math-y terminology
- [x] Follow the conventions used in the `Js.*` modules
- [ ] Type functions correctly, e.g. `compare` should return a proper variant, not `0`, `1` or `-1`
- [x] Remove operators and aliases, e.g. `pure` as an alias for `return`
- [ ] Remove the use of exceptions for internal logic
- [x] Replace `Format.printf` calls, they pull in a lot stuff for little benefit
- [ ] Document everything properly, with examples
- [ ] Add tests for everything

I think this is good enough to be merged for now

### CODE REVIEW

File: src/bstring.ml [hunk 1]
```diff
@@ -38,39 +45,78 @@ let compare = String.compare
...
+let uppercase = String.uppercase
+let escaped = String.escaped
+let trim = String.trim
 
 (*let init n f =
   let buf = Bytes.init n f in
   Bytes.unsafe_to_string buf*)
 
+external get : string -> int -> char = "%string_safe_get"
+
```
Comment (dorafmon) [inline src/bstring.ml]: Should we have an unsafe get？

File: src/bstring.ml [hunk 2]
```diff
@@ -1,13 +1,25 @@
```
Comment (dorafmon) [inline src/bstring.ml]: I see it is in TODO but I think we need unit tests.

File: src/bstring.ml (hunk 1 above)
Comment (dorafmon) [inline src/bstring.ml]: Please ignore this. Just saw it.

File: src/bstring.ml [hunk 3]
```diff
@@ -38,39 +45,78 @@ let compare = String.compare
...
+let forEachWithIndex = String.iteri
+let mapWithIndex = String.mapi
+
+let reverse s =
   let n = length s in
-  init n (fun i -> s.[n-i-1])
+  makeWithInit n (fun i -> s.[n-i-1])
 
-let rec _to_list s acc i len =
+let rec _toList s acc i len =
```
Comment (dorafmon) [inline src/bstring.ml]: I suggest we expose this API. I found it quite useful, also would parallel `toList` in `Barray`

File: src/bstring.ml (hunk 2 above)
Comment (glennsl) [inline src/bstring.ml]: We should, but the way things are moving I think we need to prioritize organization and settling on an interface first.

File: src/bstring.ml (hunk 3 above)
Comment (glennsl) [inline src/bstring.ml]: It already is, but I'm still not sure it's such a good idea. What use case would not be served better by a sequence, or some other lazy mechanism of iteration?

That said, I've been thinking about having an `Ex` module for each main module, e.g. `String.Ex` to contain functions that might be useful but have some caveat to them. This seems like a good candidate for that.

**Pull request merged**